example/contrib/socks-client/socks/impl/handshake.hpp

```diff
@@ -0,0 +1,598 @@
+                if (username_.empty())
+                {
+                    ec = socks::error::socks_username_required;
+                    break;
+                }
+
+                request.consume(request.size());
+
+                std::size_t bytes_to_write = username_.size() + password_.size() + 3;
+                auto auth = static_cast<char*>(request.prepare(bytes_to_write).data());
+
+                write<uint8_t>(0x01, auth); // auth version.
+                write<uint8_t>(static_cast<uint8_t>(username_.size()), auth);
+                std::copy(username_.begin(), username_.end(), auth);    // username.
+                auth += username_.size();
+                write<uint8_t>(static_cast<int8_t>(password_.size()), auth);
+                std::copy(password_.begin(), password_.end(), auth);    // password.
+                auth += password_.size();
+                request.commit(bytes_to_write);
+
+                // write username & password.
+                return net::async_write(
+                    stream_,
+                    request,
+                    std::move(*this));
+            }
+
+            if (method == detail::SOCKS5_AUTH_NONE) // no need auth...
+            {
+                step_ = 5;
+                return (*this)(ec, 0);
+            }
+
+            ec = socks::error::socks_unsupported_authentication_version;
+            break;
+        }
+        case 3:
+        {
+            step_ = 4;
+            response.consume(response.size());
+            return net::async_read(
+                stream_,
+                response,
+                net::transfer_exactly(2),
+                std::move(*this));
+        }
+        case 4:
+        {
+            step_ = 5;
+            auto resp = static_cast<const char*>(response.data().data());
+
+            auto version = read<uint8_t>(resp);
+            auto status = read<uint8_t>(resp);
+
+            if (version != 0x01) // auth version.
+            {
+                ec = error::socks_unsupported_authentication_version;
+                break;
+            }
+
+            if (status != 0x00)
+            {
+                ec = error::socks_authentication_error;
+                break;
+            }
+            BOOST_FALLTHROUGH;
+        }
+        case 5:
+        {
+            step_ = 6;
+
+            request.consume(request.size());
+            std::size_t bytes_to_write = 7 + hostname_.size();
+            auto req = static_cast<char*>(
+                request.prepare(std::max<std::size_t>(bytes_to_write, 22)).data());
+
+            write<uint8_t>(detail::SOCKS_VERSION_5, req); // SOCKS VERSION 5.
+            write<uint8_t>(detail::SOCKS_CMD_CONNECT, req); // CONNECT command.
+            write<uint8_t>(0, req); // reserved.
+
+            if (use_hostname_)
+            {
+                write<uint8_t>(detail::SOCKS5_ATYP_DOMAINNAME, req); // atyp, domain name.
+                BOOST_ASSERT(hostname_.size() <= 255);
```

Comment:
If protocol strings have upper limits, these need to be enforced at runtime by something stronger than just an assert. The handshaking functions need to check the parameters and throw if they are out of bounds, e.g.
```
    if (hostname.size() > 255)
        BOOST_THROW_EXCEPTION(std::out_of_range{"hostname.size() > 255"});
```